File: tools/iterm2terminal.swift
#!/usr/bin/env swift

import AppKit

enum iTermColors: String {
  case Ansi0 = "Ansi 0 Color"
  case Ansi1 = "Ansi 1 Color"
  case Ansi2 = "Ansi 2 Color"
  case Ansi3 = "Ansi 3 Color"
  case Ansi4 = "Ansi 4 Color"
  case Ansi5 = "Ansi 5 Color"
  case Ansi6 = "Ansi 6 Color"
  case Ansi7 = "Ansi 7 Color"
  case Ansi8 = "Ansi 8 Color"
  case Ansi9 = "Ansi 9 Color"
  case Ansi10 = "Ansi 10 Color"
  case Ansi11 = "Ansi 11 Color"
  case Ansi12 = "Ansi 12 Color"
  case Ansi13 = "Ansi 13 Color"
  case Ansi14 = "Ansi 14 Color"
  case Ansi15 = "Ansi 15 Color"
  case CursorText = "Cursor Text Color"
  case SelectedText = "Selected Text Color"
  case Foreground = "Foreground Color"
  case Background = "Background Color"
  case Bold = "Bold Color"
  case Selection = "Selection Color"
  case Cursor = "Cursor Color"
}

enum TerminalColors: String {
  case AnsiBlack           = "ANSIBlackColor"
  case AnsiRed             = "ANSIRedColor"
  case AnsiGreen           = "ANSIGreenColor"
  case AnsiYellow          = "ANSIYellowColor"
  case AnsiBlue            = "ANSIBlueColor"
  case AnsiMagenta         = "ANSIMagentaColor"
  case AnsiCyan            = "ANSICyanColor"
  case AnsiWhite           = "ANSIWhiteColor"
  case AnsiBrightBlack     = "ANSIBrightBlackColor"
  case AnsiBrightRed       = "ANSIBrightRedColor"
  case AnsiBrightGreen     = "ANSIBrightGreenColor"
  case AnsiBrightYellow    = "ANSIBrightYellowColor"
  case AnsiBrightBlue      = "ANSIBrightBlueColor"
  case AnsiBrightMagenta   = "ANSIBrightMagentaColor"
  case AnsiBrightCyan      = "ANSIBrightCyanColor"
  case AnsiBrightWhite     = "ANSIBrightWhiteColor"
  case Background          = "BackgroundColor"
  case Text                = "TextColor"
  case BoldText            = "TextBoldColor"
  case Selection           = "SelectionColor"
  case Cursor              = "CursorColor"
}

let iTermColor2TerminalColor = [
  iTermColors.Ansi0: TerminalColors.AnsiBlack,
  iTermColors.Ansi1: TerminalColors.AnsiRed,
  iTermColors.Ansi2: TerminalColors.AnsiGreen,
  iTermColors.Ansi3: TerminalColors.AnsiYellow,
  iTermColors.Ansi4: TerminalColors.AnsiBlue,
  iTermColors.Ansi5: TerminalColors.AnsiMagenta,
  iTermColors.Ansi6: TerminalColors.AnsiCyan,
  iTermColors.Ansi7: TerminalColors.AnsiWhite,
  iTermColors.Ansi8: TerminalColors.AnsiBrightBlack,
  iTermColors.Ansi9: TerminalColors.AnsiBrightRed,
  iTermColors.Ansi10: TerminalColors.AnsiBrightGreen,
  iTermColors.Ansi11: TerminalColors.AnsiBrightYellow,
  iTermColors.Ansi12: TerminalColors.AnsiBrightBlue,
  iTermColors.Ansi13: TerminalColors.AnsiBrightMagenta,
  iTermColors.Ansi14: TerminalColors.AnsiBrightCyan,
  iTermColors.Ansi15: TerminalColors.AnsiBrightWhite,
  iTermColors.Background: TerminalColors.Background,
  iTermColors.Foreground: TerminalColors.Text,
  iTermColors.Selection: TerminalColors.Selection,
  iTermColors.Bold: TerminalColors.BoldText,
  iTermColors.Cursor: TerminalColors.Cursor,
]

struct iTermColorComponent {
  static let Red = "Red Component"
  static let Green = "Green Component"
  static let Blue = "Blue Component"
}

func itermColorSchemeToTerminalColorScheme(_ itermColorScheme: NSDictionary, name: String) -> NSDictionary {
  var terminalColorScheme: [String: AnyObject] = [
    "name" : name as AnyObject,
    "type" : "Window Settings" as AnyObject,
    "ProfileCurrentVersion" : 2.04 as AnyObject,
    "columnCount": 90 as AnyObject,
    "rowCount": 50 as AnyObject,
    ]
  if let font = archivedFontWithName("PragmataPro", size: 14) {
    terminalColorScheme["Font"] = font as AnyObject?
  }
  for (rawKey, rawValue) in itermColorScheme {
    if let name = rawKey as? String {
      if let key = iTermColors(rawValue: name) {
        if let terminalKey = iTermColor2TerminalColor[key] {
          if let itermDict = rawValue as? [String: AnyObject] {
            let (r, g, b) = (
              floatValue(itermDict[iTermColorComponent.Red]),
              floatValue(itermDict[iTermColorComponent.Green]),
              floatValue(itermDict[iTermColorComponent.Blue]))
            let color = NSColor(deviceRed: r, green: g, blue: b, alpha: 1)
            let colorData = NSKeyedArchiver.archivedData(withRootObject: color)
            terminalColorScheme[terminalKey.rawValue] = colorData as AnyObject
          }
        }
      }
    }
  }
  return terminalColorScheme as NSDictionary
}

func archivedFontWithName(_ name: String, size: CGFloat) -> Data? {
  if let font = NSFont(name: name, size: size) {
    return NSKeyedArchiver.archivedData(withRootObject: font)
  }
  return nil
}

func floatValue(_ value: AnyObject?) -> CGFloat {
  if let num = value as? CGFloat {
    return num
  }
  return 0
}

func convertToTerminalColors(_ itermFile: String, terminalFile: NSString) {
  if let itermScheme = NSDictionary(contentsOfFile: itermFile) {
    print("converting \(itermFile) -> \(terminalFile)")
    let terminalName = ((terminalFile.lastPathComponent) as NSString).deletingPathExtension
    let terminalScheme = itermColorSchemeToTerminalColorScheme(itermScheme, name: terminalName)
    terminalScheme.write(toFile: terminalFile as String, atomically: true)
  } else {
    print("unable to load \(itermFile)")
  }
}

if CommandLine.argc > 1 {
  for itermFile in CommandLine.arguments {
    let terminalFilePath = (itermFile as NSString).deletingPathExtension + ".terminal"
    convertToTerminalColors(itermFile, terminalFile: terminalFilePath as NSString)
  }
} else {
  print("usage: iTermColorsToTerminalColors FILE.itermcolors [...]")
}

File: tools/iterm2terminal.py
#!/usr/bin/env python

# Python script to convert iTerm2 themes to Terminal.app themes
#
# Usage:
# python3 -m iterm2terminal /path/to/itermcolors/file /path/to/target/directory

import plistlib

from argparse import ArgumentParser
from pathlib import Path

RED = "Red Component"
GREEN = "Green Component"
BLUE = "Blue Component"

mappings = {
    "Ansi 0 Color": "ANSIBlackColor",
    "Ansi 1 Color": "ANSIRedColor",
    "Ansi 2 Color": "ANSIGreenColor",
    "Ansi 3 Color": "ANSIYellowColor",
    "Ansi 4 Color": "ANSIBlueColor",
    "Ansi 5 Color": "ANSIMagentaColor",
    "Ansi 6 Color": "ANSICyanColor",
    "Ansi 7 Color": "ANSIWhiteColor",
    "Ansi 8 Color": "ANSIBrightBlackColor",
    "Ansi 9 Color": "ANSIBrightRedColor",
    "Ansi 10 Color": "ANSIBrightGreenColor",
    "Ansi 11 Color": "ANSIBrightYellowColor",
    "Ansi 12 Color": "ANSIBrightBlueColor",
    "Ansi 13 Color": "ANSIBrightMagentaColor",
    "Ansi 14 Color": "ANSIBrightCyanColor",
    "Ansi 15 Color": "ANSIBrightWhiteColor",
    "Background Color": "BackgroundColor",
    "Foreground Color": "TextColor",
    "Bold Color": "TextBoldColor",
    "Selection Color": "SelectionColor",
    "Cursor Color": "CursorColor",
}


def rgb_dict_to_rgb_bytes(iterm_dict):
    r = iterm_dict[RED]
    g = iterm_dict[GREEN]
    b = iterm_dict[BLUE]

    rgb = f"{r} {g} {b}\x00"

    return bytes(rgb, "utf-8")


def create_color_entry(colors):
    rgb = rgb_dict_to_rgb_bytes(colors)
    plist = {
        "$version": 100000,
        "$objects": [
            "$null",
            {
                "NSRGB": rgb,
                "NSColorSpace": 2,
                "$class": plistlib.UID(2)
            },
            {
                "$classname": "NSColor",
                "$classes": ["NSColor", "NSObject"]
            }
        ],
        "$archiver": "NSKeyedArchiver",
        "$top": {"root": plistlib.UID(1)}
    }

    return plistlib.dumps(plist, fmt=plistlib.FMT_BINARY, sort_keys=False)


def convert(source, target):
    source_path = (Path(source)).resolve()
    if not source_path.exists():
        raise FileNotFoundError(f"Source file {source_path} does not exist")

    target_path = (Path(target)).resolve()
    if not target_path.exists():
        raise FileNotFoundError(f"Target directory {target_path} does not exist")

    target_filename = f"{source_path.stem}.terminal"
    target_path = (target_path / target_filename).resolve()

    with open(source_path, "rb") as file:
        iterm = plistlib.load(file)

    terminal = {
        "name": source_path.stem,
        "type": "Window Settings",
        "ProfileCurrentVersion": 2.04,
        "columnCount": 90,
        "rowCount": 50,
    }

    for key, value in iterm.items():
        if key in mappings:
            terminal.update({mappings.get(key): create_color_entry(value)})

    with open(target_path, "wb") as output_file:
        plistlib.dump(terminal, output_file)


if __name__ == "__main__":
    parser = ArgumentParser()
    parser.add_argument("source", help="Location of .itermcolors file or directory to convert")
    parser.add_argument("target", help="Directory in which to save Terminal.app color scheme")
    args = parser.parse_args()

    source_path = Path(args.source).resolve()
    target_path = Path(args.target).resolve()

    print(" * Generating Terminal.app color scheme(s)...")

    if not target_path.exists():
         print(f" * Target directory {target_path} does not exist. Creating it...")
         target_path.mkdir(parents=True, exist_ok=True)
    else:
         print(f" * Using existing target directory: {target_path}")

    def process_file(file_path, target_dir):
        #print(f"Processing file: {file_path}")
        try:
            convert(str(file_path), str(target_dir))
            #print(f"Successfully converted {file_path} to {target_dir}")
        except FileNotFoundError as e:
            print(e)

    if source_path.is_dir():
        print(f" * Source is a directory: {source_path}")
        files = list(source_path.glob("*.itermcolors"))
        if not files:
            print(f"No .itermcolors files found in {source_path}")
        for iterm_file in files:
            process_file(iterm_file, target_path)
    elif source_path.is_file():
        print(f" *Source is a file: {source_path}")
        process_file(source_path, target_path)
    else:
        print(f"Source {source_path} does not exist")
        exit(1)
